File: Python/textbook_track/chapter08/ba8d.py
# Implement the Soft k-Means Clustering Algorithm
import math
import sys

NR_ITERATIONS = 100


def convert_to_numbers(line, converter_func):
    return [converter_func(elem) for elem in line.split()]


def read_points(reader):
    return [convert_to_numbers(line, float) for line in reader]


def euclidean_distance(p, q):
    return math.sqrt(sum((p_i - q_i)**2 for p_i, q_i in zip(p, q)))


def assign_points_to_centers(points, centers, beta, k):
    n = len(points)
    responsibility_matrix = [[0.0] * n for _ in range(k)]
    for ix, center in enumerate(centers):
        for jy, point in enumerate(points):
            dist = euclidean_distance(point, center)
            responsibility_matrix[ix][jy] = math.exp(-beta * dist)
    normalize_responsibility_matrix(responsibility_matrix)
    return responsibility_matrix


def normalize_responsibility_matrix(responsibility_matrix):
    column_sums = [sum(column) for column in zip(*responsibility_matrix)]
    for ix, row in enumerate(responsibility_matrix):
        for jy, (elem, column_sum) in enumerate(zip(row, column_sums)):
            responsibility_matrix[ix][jy] = elem / column_sum


def dot_product(v, w):
    return sum(v_i * w_i for v_i, w_i in zip(v, w))


def recalculate_centers(points, responsibility_matrix, k, m):
    centers = [[0.0] * m for _ in range(k)]
    row_sums = [sum(row) for row in responsibility_matrix]

    for jy, data_column in enumerate(zip(*points)):
        for ix, (responsibility_row, row_sum) in enumerate(zip(responsibility_matrix, row_sums)):
            centers[ix][jy] = dot_product(responsibility_row, data_column) / row_sum
    return centers


def run_soft_k_means_clustering(points, k, m, beta):
    centers = points[:k]
    for _ in range(NR_ITERATIONS):
        responsibility_matrix = assign_points_to_centers(points, centers, beta, k)
        centers = recalculate_centers(points, responsibility_matrix, k, m)
    return centers


def main():
    reader = sys.stdin
    k, m = convert_to_numbers(next(reader), int)
    beta = float(next(reader))
    points = read_points(reader)
    result = run_soft_k_means_clustering(points, k, m, beta)
    for point in result:
        print(' '.join(map(str, point)))


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter08/BA8D.scala
package TextbookTrack.Chapter08

object BA8D {
  import scala.reflect.ClassTag
  import scala.annotation.tailrec

  type Point = List[Double]

  private val NumberOfIterations: Int = 100

  private def convertToNumber[T: ClassTag](line: String, converter: String => T): List[T] =
    line.split(" ").map(converter).toList

  private def readParameters(line: String): (Int, Int) = convertToNumber(line, _.toInt) match {
    case List(k, m) => (k, m)
    case _ => throw new Exception("Unexpected number of parameters.")
  }

  private def euclideanDistance(p: Point, q: Point): Double =
    math.sqrt(p.lazyZip(q).foldLeft(0.0){ case (d, (pi, qi)) => d + math.pow(pi - qi, 2) })

  private def assignPointsToCenters(points: List[Point], centers: List[Point], beta: Double): List[List[Double]] = {
    val responsibilityMatrix: List[List[Double]] =
      centers.map{ center => points.map(point => math.exp(-beta * euclideanDistance(center, point))) }
    val columnSums: List[Double] = responsibilityMatrix.transpose.map(_.sum)
    responsibilityMatrix.map(row => row.zip(columnSums).map{ case (elem, sum) => elem / sum })
  }

  private def dotProduct(v: List[Double], w: List[Double]): Double =
    v.lazyZip(w).foldLeft(0.0){ case (s, (a, b)) => s + a * b }

  private def recalculateCenters(points: List[Point], responsibilityMatrix: List[List[Double]]): List[Point] = {
    val dataColumns: List[List[Double]] = points.transpose
    val rowSums: List[Double] = responsibilityMatrix.map(_.sum)
    responsibilityMatrix
      .lazyZip(rowSums)
      .map{ case (responsibilityRow, sum) => dataColumns.map(dotProduct(responsibilityRow, _) / sum) }
  }

  def runSoftKMeansClustering(points: List[Point], k: Int, beta: Double): List[Point] = {
    @tailrec
    def loop(centers: List[Point], nrIterations: Int): List[Point] = {
      if (nrIterations == NumberOfIterations) centers
      else {
        val responsibilityMatrix: List[List[Double]] = assignPointsToCenters(points, centers, beta)
        val updatedCenters: List[Point] = recalculateCenters(points, responsibilityMatrix)
        loop(updatedCenters, nrIterations + 1)
      }
    }

    loop(points.take(k), 0)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val (k, _): (Int, Int) = readParameters(reader.next())
    val beta: Double = reader.next().toDouble
    val points: List[Point] = reader.map(convertToNumber(_, _.toDouble)).toList
    val result: List[Point] = runSoftKMeansClustering(points, k, beta)
    result.foreach(point => println(point.mkString(" ")))
  }
}
